File: Android/FoodRescue/app/src/main/java/edu/ewu/team1/foodrescue/fragments/FeederFragment.java
package edu.ewu.team1.foodrescue.fragments;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.TextView;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.LatLng;

import java.math.BigInteger;
import java.util.HashMap;
import java.util.Map;
import java.util.Random;

import edu.ewu.team1.foodrescue.R;
import edu.ewu.team1.foodrescue.cryptography.DiffieHelmanEncryptedMessage;

/**
 * A simple {@link Fragment} subclass.
 */
public class FeederFragment extends Fragment implements OnMapReadyCallback {
    //https://developers.google.com/maps/documentation/android-api/groundoverlay
    private MapView mapView;
    private GoogleMap map;
    private String[] names;
    private Double[] lats;
    private Double[] lngs;
    private Spinner spinner;
    private boolean hasLocationAccess = false;
    private final int gpsPermissionsRequestCode = 456;
    private View view;
    private EditText editText;
    private TextView location;

    public FeederFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_feeder, container, false);

        location = view.findViewById(R.id.textViewLatLng);
        populateDropdownMenu();

        //Map
        mapView = view.findViewById(R.id.mapView);
        mapView.onCreate(savedInstanceState);
        mapView.getMapAsync(this);

        //Define behavior of "Send Announcement" button
        editText = view.findViewById(R.id.editTextMessage);
        Button buttonSubmit = view.findViewById(R.id.buttonSubmit);
        buttonSubmit.setOnClickListener(view -> {
            LatLng loc = getCrosshairLocation();
            assert loc != null;
            String locName = names[spinner.getSelectedItemPosition()];
            String message = editText.getText().toString();

            RequestQueue queue = Volley.newRequestQueue(getContext());
            String url = "bradstephensoncode.org/FoodRescue/sendrest.php";
            StringRequest postRequest = new StringRequest(Request.Method.POST, url,
                    response -> Log.d("Response", response),
                    error -> Log.d("Error.Response", error.getMessage())
            ) {
                @Override
                protected Map<String, String> getParams() {
                    Map<String, String> params = new HashMap<>();
                    //TODO: Include hashed username here
                    params.put("lat", String.valueOf(loc.latitude));
                    params.put("lng", String.valueOf(loc.longitude));
                    params.put("locName", locName);
                    params.put("message", message);

                    return params;
                }
            };
            queue.add(postRequest);

        });

        return view;
    }

    /**
     * When the map is loaded we configure some settings for it
     *
     * @param googleMap the map which is now ready to be configured and used
     */
    @Override
    public void onMapReady(GoogleMap googleMap) {
        map = googleMap;
        moveMapToLocation(getCurrentLocation());
        displayCurrentLocation();
        map.setOnCameraMoveListener(this::displayCurrentLocation);
    }

    /**
     * Displays the LatLng of the center of the crosshairs in a text view
     */
    private void displayCurrentLocation() {
        LatLng loc = getCrosshairLocation();
        location.setText(loc.toString());
    }

    /**
     * Callback for requesting permissions
     */
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        if (requestCode == gpsPermissionsRequestCode) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
                hasLocationAccess = true;
            }
        }
    }

    /**
     * Gets the users current location as accurately as possible
     *
     * @return LatLng
     */
    @SuppressLint("MissingPermission")
    private LatLng getCurrentLocation() {
        //https://stackoverflow.com/questions/20210565/android-location-manager-get-gps-location-if-no-gps-then-get-to-network-provid
        LatLng latLng;
        if (hasLocationAccess) {
            boolean gps_enabled;
            boolean network_enabled;

            LocationManager lm = (LocationManager) getActivity().getSystemService(Context.LOCATION_SERVICE);

            assert lm != null;
            gps_enabled = lm.isProviderEnabled(LocationManager.GPS_PROVIDER);
            network_enabled = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER);

            Location net_loc = null, gps_loc = null, finalLoc = null;

            if (gps_enabled)
                gps_loc = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER);
            if (network_enabled)
                net_loc = lm.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);

            if (gps_loc != null && net_loc != null) {

                //smaller the number more accurate result will be
                if (gps_loc.getAccuracy() > net_loc.getAccuracy())
                    finalLoc = net_loc;
                else
                    finalLoc = gps_loc;

            } else {

                if (gps_loc != null) {
                    finalLoc = gps_loc;
                } else if (net_loc != null) {
                    finalLoc = net_loc;
                }
            }
            if (finalLoc == null) {
                latLng = new LatLng(47.491376, -117.582917);
            } else {
                latLng = new LatLng(finalLoc.getLatitude(), finalLoc.getLongitude());
            }
        }
        //TODO: Comment this out. It is for testing purposes
        //This way, the location will be centered on campus instead of your house
        latLng = new LatLng(47.491376, -117.582917);

        return latLng;
    }

    /**
     * Gets the LatLng of the center of the map
     *
     * @return LatLng
     */
    private LatLng getCrosshairLocation() {
        return map.getCameraPosition().target;
    }


    /**
     * Centers the map on the specified location and resets the zoom level
     *
     * @param latLng The location to center the map on
     */
    @SuppressLint("MissingPermission")
    private void moveMapToLocation(LatLng latLng) {
        if (hasLocationAccess) map.setMyLocationEnabled(true);
        map.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 17.0f));
    }

    /**
     * Populates the dropdown menu with all the buildings and their locations
     * Closest stored location to the user is duplicated and put at the top of the list
     */
    private void populateDropdownMenu() {
        //Make sure we have permission to get the users location
        //The users location will be used to order the dropdown list
        if (ActivityCompat.checkSelfPermission(getActivity(),
                Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(getActivity(),
                        Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(getActivity(), new String[]{Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION}, gpsPermissionsRequestCode);
        }

        //Populating the arrays we will use
        names = getResources().getStringArray(R.array.locationNames);
        lats = new Double[names.length + 1];
        lngs = new Double[names.length + 1];

        int i = 1;
        for (String c : getResources().getStringArray(R.array.locationLats)) {
            lats[i++] = Double.parseDouble(c);
        }
        i = 1;
        for (String c : getResources().getStringArray(R.array.locationLngs)) {
            lngs[i++] = Double.parseDouble(c);
        }

        //Copy the nearest stored location to the user to the top of the list
        LatLng currLoc = getCurrentLocation();
        double smallest = Double.MAX_VALUE;
        int closest = 1;
        for (i = 2; i < names.length; i++) {
            double dist = Math.hypot(lats[i] - currLoc.latitude, lngs[i] - currLoc.longitude);
            if (dist < smallest) {
                closest = i;
                smallest = dist;
            }
        }
        names[0] = names[closest];
        lats[0] = lats[closest];
        lngs[0] = lngs[closest];

        //Populate menu
        spinner = view.findViewById(R.id.spinner);
        spinner.setAdapter(new ArrayAdapter<>(getActivity(), android.R.layout.simple_list_item_1, names));
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> parentView, View selectedItemView, int position, long id) {
                // your code here
                if (position != 1)//Position 1 is "Other"
                    moveMapToLocation(new LatLng(lats[position], lngs[position]));
            }

            @Override
            public void onNothingSelected(AdapterView<?> parentView) {
                // your code here
            }

        });
    }

    @Override
    public void onResume() {
        mapView.onResume();
        super.onResume();
    }


    @Override
    public void onPause() {
        super.onPause();
        mapView.onPause();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        mapView.onDestroy();
    }

    @Override
    public void onLowMemory() {
        super.onLowMemory();
        mapView.onLowMemory();
    }
}

File: Android/FoodRescue/app/src/main/java/edu/ewu/team1/foodrescue/fragments/FeederFragment.kt
package edu.ewu.team1.foodrescue.fragments

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.android.volley.Response
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.model.LatLng
import edu.ewu.team1.foodrescue.MainActivity
import edu.ewu.team1.foodrescue.R
import edu.ewu.team1.foodrescue.notifications.NotificationShower
import edu.ewu.team1.foodrescue.utilities.ConfirmDialog
import edu.ewu.team1.foodrescue.utilities.DataManager
import edu.ewu.team1.foodrescue.utilities.FoodEvent
import edu.ewu.team1.foodrescue.utilities.VolleyWrapper
import java.util.*

class FeederFragment : Fragment() {
	//https://developers.google.com/maps/documentation/android-api/groundoverlay
	private lateinit var mapView: MapView
	private var map: GoogleMap? = null
	private lateinit var names: Array<String>
	private lateinit var spinnerLocation: Spinner
	private lateinit var spinnerExpiry: Spinner
	private var hasLocationAccess = false
	private lateinit var location: TextView
	private lateinit var dataManager: DataManager
	private lateinit var editText: EditText

	private val crosshairLocation: LatLng?
		get() = map?.cameraPosition?.target

	companion object {
		const val gpsPermissionsRequestCode = 456
		const val CUSTOM_MESSAGE = "MESSAGE"
		const val SELECTED_LOCATION = "LOCATION"
		const val SELECTED_EXPIRY = "EXPIRY"
		const val LAT = "LAT"
		const val LNG = "LNG"
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		super.onCreateView(inflater, container, savedInstanceState)
		Log.e("onCreateView()", "called")
		retainInstance
		return inflater.inflate(R.layout.fragment_feeder, container, false)
	}

	override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
		Log.e("onViewCreated()", "called")
		dataManager = (activity as MainActivity).dataManager
		location = view.findViewById(R.id.textViewLatLng)
		editText = view.findViewById(R.id.editTextMessage)
		loadMap(view, savedInstanceState)
		populateLocationMenu(view)
		populateExpiryMenu(view)
		defineSubmitButtonBehavior(view)
		defineShowButtonBehavior(view)
		//Restore instance state
		if (savedInstanceState != null) {
			val msg = savedInstanceState.getString(CUSTOM_MESSAGE)
			editText.setText(msg.toCharArray(), 0, msg.length)
			spinnerLocation.setSelection(savedInstanceState.getInt(SELECTED_LOCATION))
			spinnerExpiry.setSelection(savedInstanceState.getInt(SELECTED_EXPIRY))
		}
	}

	override fun onSaveInstanceState(outState: Bundle) {
		Log.e("onSaveInstanceState()", "called")
		super.onSaveInstanceState(outState)
		outState.putString(CUSTOM_MESSAGE, editText.text.toString())
		outState.putInt(SELECTED_LOCATION, spinnerLocation.selectedItemPosition)
		outState.putInt(SELECTED_EXPIRY, spinnerExpiry.selectedItemPosition)
		outState.putDouble(LAT, crosshairLocation?.latitude ?: 47.491355)
		outState.putDouble(LNG, crosshairLocation?.longitude?: -117.582798)
	}

	//<editor-fold desc="onCreate() Helper Methods Section">
	/**
	 * Populates the dropdown menu with all the buildings and their locations
	 * Closest stored location to the user is duplicated and put at the top of the list
	 */
	private fun populateLocationMenu(view: View) {
		//Make sure we have permission to get the users location
		//The users location will be used to order the dropdown list
		if (ActivityCompat.checkSelfPermission(activity!!,
						Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(activity!!,
						Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
			ActivityCompat.requestPermissions(activity!!, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), gpsPermissionsRequestCode)
		}

		//Populating the arrays we will use
		names = resources.getStringArray(R.array.locationNames)
		val lngs = resources.getStringArray(R.array.locationLats).iterator()
		val lats = resources.getStringArray(R.array.locationLngs).iterator()
		val latitudes = Array(names.size, {
			lngs.next().toDouble()
		})
		val longitudes = Array(names.size, {
			lats.next().toDouble()
		})

		//Copy the nearest stored location to the user to the top of the list
		val currLoc = currentLocation
		var smallest = java.lang.Double.MAX_VALUE
		var closest = 1
		var i = 2
		while (i < names.size) {
			val dist = Math.hypot(latitudes[i] - currLoc.latitude, longitudes[i] - currLoc.longitude)
			if (dist < smallest) {
				closest = i
				smallest = dist
			}
			i++
		}
		names[0] = names[closest]
		latitudes[0] = latitudes[closest]
		longitudes[0] = longitudes[closest]

		//Populate menu
		spinnerLocation = view.findViewById(R.id.spinnerLocation)
		spinnerLocation.adapter = ArrayAdapter(activity!!, android.R.layout.simple_list_item_1, names)
		spinnerLocation.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
			override fun onItemSelected(parentView: AdapterView<*>, selectedItemView: View?, position: Int, id: Long) {
				Log.e("onItemSelected()", "called")
				//Position 1 is "Other"
				if (position != 1)
					moveMapToLocation(LatLng(latitudes[position], longitudes[position]))
			}

			override fun onNothingSelected(parentView: AdapterView<*>) {}
		}
	}

	private fun populateExpiryMenu(view: View) {
		spinnerExpiry = view.findViewById<Spinner>(R.id.spinnerExpiry)
		spinnerExpiry.adapter = ArrayAdapter(activity!!, android.R.layout.simple_list_item_1, arrayOf("<15 minutes", "<30 minutes", "<1 hour"))
		spinnerExpiry.setSelection(0)
	}

	private fun defineSubmitButtonBehavior(view: View) {
		//Define behavior of "Send Announcement" button
		view.findViewById<Button>(R.id.buttonSubmit).setOnClickListener {
			ConfirmDialog.confirmAction(Runnable {
				val editTextMessage = view.findViewById<EditText>(R.id.editTextMessage)
				val loc = crosshairLocation
				val locName = names[spinnerLocation.selectedItemPosition]
				val message = editTextMessage.text.toString()
				val duration = when (view.findViewById<Spinner>(R.id.spinnerExpiry).selectedItemPosition) {
					0 -> 15
					1 -> 30
					else -> 60
				}
				val url = MainActivity.SERVER_IP + MainActivity.SEND_NOTIFICATION

				val params = HashMap<String, String>()
				val token = dataManager.getToken()
				params["title"] = locName
				params["body"] = message
				params["lat"] = loc!!.latitude.toString()
				params["lng"] = loc.longitude.toString()
				params["expiry"] = duration.toString()

				VolleyWrapper.post(view.context, url, params, Response.Listener { response ->
					Log.e("post Response:", response)
					//TODO: Check if the server allowed the notification to be sent
					//If it didn't, the user should be notified of how to become an authorized feeder
					Toast.makeText(view.context, "Notification sent!", Toast.LENGTH_LONG).show()
				})
				editTextMessage.text.clear()
			}, R.string.confirm_send, context!!)

		}
	}

	private fun defineShowButtonBehavior(view: View) {
		view.findViewById<Button>(R.id.buttonShowLocal).setOnClickListener {
			ConfirmDialog.confirmAction(Runnable {
				val loc = crosshairLocation
				val locName = names[spinnerLocation.selectedItemPosition]
				val message = view.findViewById<EditText>(R.id.editTextMessage).text.toString()
				val duration = when (view.findViewById<Spinner>(R.id.spinnerExpiry).selectedItemPosition) {
					0 -> 15
					1 -> 30
					else -> 60
				}
				NotificationShower.show(FoodEvent(locName, message, loc?.latitude, loc?.longitude, duration, System.currentTimeMillis()), dataManager, view.context)
			}, R.string.confirm_send, context!!)
		}
	}

	private fun loadMap(view: View, savedInstanceState: Bundle?) {
		mapView = view.findViewById(R.id.mapView)
		mapView.onCreate(savedInstanceState)
		mapView.getMapAsync({
			map = it
			map!!.setOnCameraMoveListener({ this.displayCurrentLocation() })

			if (savedInstanceState != null) {
				moveMapToLocation(LatLng(
						savedInstanceState.getDouble(LAT),
						savedInstanceState.getDouble(LNG)
				))
			} else {
				moveMapToLocation(currentLocation)
			}
			displayCurrentLocation()
		})
	}
	//</editor-fold>

	//<editor-fold desc="Map Helper Methods Section">
	/**
	 * Gets the users current location as accurately as possible
	 * https://stackoverflow.com/questions/20210565/android-location-manager-get-gps-location-if-no-gps-then-get-to-network-provid
	 * @return LatLng
	 */
	private val currentLocation: LatLng
		@SuppressLint("MissingPermission")
		get() {
			var latLng = LatLng(47.491376, -117.582917)
			if (hasLocationAccess) {
				val lm = activity!!.getSystemService(Context.LOCATION_SERVICE) as LocationManager
				val isGpsEnabled = lm.isProviderEnabled(LocationManager.GPS_PROVIDER)
				val isNetworkEnabled = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER)

				var gpsLocation: Location? = null
				var networkLocation: Location? = null

				if (isGpsEnabled)
					gpsLocation = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER)
				if (isNetworkEnabled)
					networkLocation = lm.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)

				if (gpsLocation != null && networkLocation != null) {
					//smaller the number more accurate result will be
					val finalLoc =
							if (gpsLocation.accuracy > networkLocation.accuracy)
								networkLocation
							else
								gpsLocation
					latLng = LatLng(finalLoc.latitude, finalLoc.longitude)
				}
			}

			return latLng
		}

	/**
	 * Displays the LatLng of the center of the crosshairs in a text view
	 */
	private fun displayCurrentLocation() {
		val loc = crosshairLocation
		location.text = loc.toString()
	}

	/**
	 * Centers the map on the specified location and resets the zoom level
	 *
	 * @param latLng The location to center the map on
	 */
	@SuppressLint("MissingPermission")
	private fun moveMapToLocation(latLng: LatLng) {
		if(map != null) {
			if (hasLocationAccess) map!!.isMyLocationEnabled = true
			map!!.moveCamera(CameraUpdateFactory.newLatLngZoom(latLng, 17f))
		}
	}
	//</editor-fold>

	//<editor-fold desc="Map Callbacks">
	/**
	 * When the map is loaded we configure some settings for it
	 *
	 * @param googleMap the map which is now ready to be configured and used
	 */

	/**
	 * Callback for requesting permissions
	 */
	override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
		if (requestCode == gpsPermissionsRequestCode) {
			if (grantResults[0] == PackageManager.PERMISSION_GRANTED && grantResults[1] == PackageManager.PERMISSION_GRANTED) {
				hasLocationAccess = true
			}
		}
	}

	override fun onResume() {
		mapView.onResume()
		super.onResume()
	}


	override fun onPause() {
		super.onPause()
		mapView.onPause()
	}

	override fun onDestroy() {
		super.onDestroy()
		mapView.onDestroy()
	}

	override fun onLowMemory() {
		super.onLowMemory()
		mapView.onLowMemory()
	}
	//</editor-fold>
}
